packages/agent/src/cbor.ts

```diff
@@ -1,126 +1,46 @@
-  public match(value: any): boolean {
-    return value && value._isPrincipal === true;
-  }
-
-  public encode(v: Principal): cbor.CborValue {
-    return cbor.value.bytes(uint8ToBuf(v.toUint8Array()));
-  }
-}
-
-class BufferEncoder implements CborEncoder<ArrayBuffer> {
-  public get name() {
-    return 'Buffer';
-  }
-
-  public get priority() {
-    return 1;
-  }
-
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  public match(value: any): boolean {
-    return value instanceof ArrayBuffer || ArrayBuffer.isView(value);
-  }
-
-  public encode(v: ArrayBuffer): cbor.CborValue {
-    return cbor.value.bytes(v);
-  }
-}
-
-class BigIntEncoder implements CborEncoder<bigint> {
-  public get name() {
-    return 'BigInt';
-  }
-
-  public get priority() {
-    return 1;
-  }
-
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  public match(value: any): boolean {
-    return typeof value === `bigint`;
-  }
-
-  public encode(v: bigint): cbor.CborValue {
-    // Always use a bigint encoding.
-    if (v > BigInt(0)) {
-      return cbor.value.tagged(2, cbor.value.bytes(uint8ToBuf(hexToBytes(v.toString(16)))));
-    } else {
-      return cbor.value.tagged(
-        3,
-        cbor.value.bytes(uint8ToBuf(hexToBytes((BigInt('-1') * v).toString(16)))),
-      );
-    }
-  }
+/**
+ * Used to extend classes that need to provide a custom value for the CBOR encoding process.
+ */
+export abstract class ToCborValue {
+  /**
+   * Returns a value that can be encoded with CBOR. Typically called in the replacer function of the {@link encode} function.
+   */
+  public abstract toCborValue(): cbor.CborValue;
 }
 
-const serializer = SelfDescribeCborSerializer.withDefaultEncoders(true);
-serializer.addEncoder(new PrincipalEncoder());
-serializer.addEncoder(new BufferEncoder());
-serializer.addEncoder(new BigIntEncoder());
-
-export enum CborTag {
-  Uint64LittleEndian = 71,
-  Semantic = 55799,
+function hasCborValueMethod(value: unknown): value is ToCborValue {
+  return typeof value === 'object' && value !== null && 'toCborValue' in value;
 }
 
 /**
- * Encode a JavaScript value into CBOR.
+ * Encode a JavaScript value into CBOR. If the value is an instance of {@link ToCborValue},
+ * the {@link ToCborValue.toCborValue} method will be called to get the value to encode.
  * @param value The value to encode
  */
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-export function encode(value: any): Uint8Array {
-  return uint8FromBufLike(serializer.serialize(value));
-}
-
-function decodePositiveBigInt(buf: Uint8Array): bigint {
-  const len = buf.byteLength;
-  let res = BigInt(0);
-  for (let i = 0; i < len; i++) {
-    res = res * BigInt(0x100) + BigInt(buf[i]);
-  }
-
-  return res;
-}
+export function encode(value: unknown): Uint8Array {
+  try {
+    return cbor.encodeWithSelfDescribedTag(value, value => {
+      if (value instanceof Principal) {
```

Comment:
Here I think we should have a more sophisticated check - the principal may have been passed from a different version of the library, so we can lean on the `_isPrincipal` tag. Same goes for Expiry below